### PR TITLE
Feature: automatic history management

### LINKED ISSUE #1595: Make possible to customize themes "history" behavior consistently
Bash-it base theme provides `_save-and-reload-history()` function that does history actions depending on user settings.

While this is defined as `[[ $autosave -eq 1 ]] && history -a && history -c && history -r `, the input parameter isn't a global bash-it variable, and so each theme may use its own (the only one supporting it, powerline, seems to use `$HISTORY_AUTOSAVE`.

In the other side, I see various themes doing:

```bash
prompt_setter() {
  local exit_status=$?

  # Save history
  history -a
  history -c
  history -r

  PS1="foooo"
}
```
Now, I've a big history file, so while I do want to use `history -a` all the times, I don't want to reload it once saved, but do it on-demand.

At the same time, some themes don't do this, so there's no consistency.

So, things to do are:
 - Provide `HISTORY_AUTOSAVE_*` variables to define what actions to do with history
 - Use `_save-and-reload-history()` in all the themes

### PR DESCRIPTION
**Pull request by gaelicWizard**
## Description
Two new functions `_bash_it_history_auto_save()` and `_bash_it_history_auto_load()`, which append new history to disk and load new history from disk, respectively.

These functions operate based on four modes configured in `$HISTCONTROL`:
1. `autoshare`: append to `$HISTFILE` when `_bash_it_history_auto_save()` called, and read from `$HISTFILE` when `_bash_it_history_auto_load()` called.
2. `autosave`: append to `$HISTFILE` when `_bash_it_history_auto_save()` called, but do not do anything when `_bash_it_history_auto_load()` called.
3. `autoload`: do not do anything when `_bash_it_history_auto_save()` called, but do read from `$HISTFILE` when `_bash_it_history_auto_load()` called.
4. `noauto` (default): don't do anything.

## Motivation and Context
This PR is based on @davidpfarrell's suggestions in and fixes #1595. 

`_bash_it_history_auto_save()` should be added to `preexec` hook and `_bash_it_history_auto_load()` should be added to `precmd` hook, so history is saved immediately and loaded on the next prompt.

Ok, I've refactored slightly. There's four commits. The first three just do a little cleanup on the existing plugins, and the last one adds the "automatic history management" feature. 

A few minor notes:
- Don't use `export` as it exports the variable to the environment of external binaries, which both adds clutter and alsö can *occasionally* (`ksh`?) cause very unexpected weirds including potential loss of the history file. We don't want a newly executed `bash` or `ksh` to inherit `$HISTFILE` but then otherwise set `$HISTFILESIZE` to anything shorter. `readonly` only survives the *current* shell, so make sure not to `export`.
- Do use `readonly` in `plugin/history-eternal` so that all three variables (`$HISTSIZE`, `$HISTFILESIZE`, and `$HISTFILE`) are all set simultaneously and aren't mis-matched by some clumsy script or plugin somewhere else. This combination ensures that the user's eternal history file is only touched with proper settings configured (or by hand).
- The two new history management functions *both* save history (`history -a`) on every run (unless configured not to). This is due to the fact that the `preexec` hook sometimes doesn't run due to weirds relating to emulating using the `DEBUG` hook, so just I save history twice. The cases where the `preexec` hook doesn't run are nearly instantaneous, so no functionality is lost. This doesn't have a performance penalty as _Bash_ doesn't actually save twice; it doesn't do anything when asked to append to an up-to-date history file. This combination maximizes the synchronicity of multiple shells writing to the file. Both running shells should see and update the file immediately, before and after user input.

I don't think that history management should be handled by the theme. We can set some defaults in `plugins/history` and let the user customize further. Maybe add to the template?

## How Has This Been Tested?
Tested locally, and all tests pass.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] If I have added a new file, I also added it to ``clean_files.txt`` and formatted it using ``lint_clean_files.sh``.
- [ ] I have added tests to cover my changes, and all the new and existing tests pass.

### CODE REVIEW

**Review by NoahGorny — APPROVED** [review]
Seems like a good idea, I'll let other folks chime in :smile:

**Comment by NoahGorny** [comment]
@davidpfarrell wanna take a look before I merge?

**Comment by cornfeedhobo** [comment]
@gaelicWizard so far I like this, but do want to do a deeper review after preexec has been dealt with.

As for right now, I'd also like to add "infinite history", but not sure if that belongs in a dedicated plugin.

```bash
export HISTFILE=$HOME/.bash_history_unlimited
export HISTSIZE=
export HISTFILESIZE=
```

**Comment by gaelicWizard** [comment]
(No need for `export`)

File: themes/base.theme.bash
Comment (gaelicWizard) [reply themes/base.theme.bash:623]: I'd like to remove this function entirely, but I don't want to break anything in between now and then. That's why it's kinda shoehorned in there like that...

**Comment by gaelicWizard** [comment]
> I've read through the linked issue, and I'm kinda stuck on the question of why we are letting themes mess with history at all. I looked through the repo's various `prompt_setter` functions and I don't see critical reasons for them to be doing that at all. Can someone help me understand?

I agree completely; my goal here is to fix/reduce/remove themes from managing history.

File: themes/base.theme.bash
Comment (gaelicWizard) [reply themes/base.theme.bash:623]: @cornfeedhobo, let me know if you want me to specifically fix up this function to be less funky or if it's OK to leave it be since I plan to delete it in short order. (My PR to remove it is already up and marked draft pending this PR; see #1951.)

**Comment by gaelicWizard** [comment]
I've rebased this on the recent cleanup of the history plugins, but I'm going to test it out for a minute to make sure ń̥o҉̙̻̖̲̮͓̩̀t̵̬̳͔ḩ̡̪͇͈i̼̦̮̙͝n̸̸̲̲͍̟g͏̭͖̠̩̭ went funny in the mean time. 😃

**Comment by gaelicWizard** [comment]
Ok, I've refactored slightly. There's four commits. The first two just do a little cleanup on the existing plugins, and the second two add the "automatic history management" feature. 

A few minor notes:
- Don't use `export` as it exports the variable to the environment of external binaries, which both adds clutter and alsö can *occasionally* (`ksh`?) cause very unexpected weirds including potential loss of the history file. We don't want a newly executed `bash` or `ksh` to inherit `$HISTFILE` but then otherwise set `$HISTFILESIZE` to anything shorter. `readonly` only survives the *current* shell, so make sure not to `export`.
- Do use `readonly` in `plugin/history-eternal` so that all three variables (`$HISTSIZE`, `$HISTFILESIZE`, and `$HISTFILE`) are all set simultaneously and aren't mis-matched by some clumsy script or plugin somewhere else. This combination ensures that the user's eternal history file is only touched with proper settings configured (or by hand).
- New history saving functions *both* save history (`history -a`) on every run (unless configured not to). The `preexec` version sometimes doesn't run due to weirds with emulating using the `DEBUG` hook, so just save it twice. This doesn't have a performance penalty as _Bash_ doesn't actually save twice; it doesn't do anything when asked to append to an up-to-date history file. This combination maximizes the synchronicity of multiple shells writing to the file. Both running shells should see and update the file immediately, before and after user input.

**Comment by gaelicWizard** [comment]
At long last, I'm happy to report that this bad boy is ready to go!

**Pull request merged**
